query/src/main/scala/filodb/query/exec/rangefn/RateFunctions.scala

```diff
@@ -49,6 +49,71 @@ object RateFunctions {
       Some(result)
     }
   }
+
+  def derivFunction(window: Window): Option[Double] = {
+    if (window.size < 2) {
+      None  // cannot calculate result without 2 samples
+    } else {
+      // We pass in an arbitrary timestamp that is near the values in use
+      // to avoid floating point accuracy issues
+      Some(linearRegression(window, window(0).timestamp)._1)
+    }
+  }
+
+  /**
+    * Logic is kept consistent with Prometheus' linearRegression function in order to get consistent results.
+    * We can look at optimizations (if any) later.
+    */
+  def linearRegression(window: Window, interceptTime: Long): Tuple2[Double, Double] = {
+    var n = 0.0
+    var sumX = 0.0
+    var sumY = 0.0
+    var sumXY = 0.0
+    var sumX2 = 0.0
+    for (i <- 0 until window.size) {
+      var sample = window(i)
+      var x = (sample.timestamp-interceptTime).toDouble / 1000.0
+      n += 1.0
+      sumY += sample.value
+      sumX += x
+      sumXY += x * sample.value
+      sumX2 += x * x
+    }
+    val covXY = sumXY - sumX*sumY/n
+    val varX = sumX2 - sumX*sumX/n
+    val slope = covXY / varX
+    val intercept = sumY/n - slope*sumX/n
+    Tuple2 (slope, intercept)
+  }
+
+  def instantValue(startTimestamp: Long,
+                       endTimestamp: Long,
+                       window: Window,
+                       isRate: Boolean): Option[Double] = {
+    if (window.size < 2) {
+      None  // cannot calculate result without 2 samples
```

Comment:
Is this an exception? Or perhaps NaN?

Reply:
This is not an exception. This essentially indicates that the time window is too small to calculate any rate. This should ideally result in a NoValue (NA or null) but we do not have such a capability in our result vectors. Hence we use an NaN. If the Option is removed and brought up into the method, this would be replaced with NaN